docs/design.md

```diff
@@ -0,0 +1,299 @@
+> in order from earliest to head.
+
+An Advertisement has several properties (see
+[the full spec](https://github.com/ipni/specs/blob/90648bca4749ef912b2d18f221514bc26b5bef0a/IPNI.md#advertisements)),
+we need the following ones:
+
+- **`PreviousID`** is the CID of the previous advertisement, and is empty for
+  the 'genesis'.
+
+- **`Metadata`** represents additional opaque data. The metadata for Graphsync
+  retrievals includes the PieceCID that we are looking for.
+
+- **`Entries`** provide a paginated list of multihashes. For our purposes, it's
+  enough to take the first entry and ignore the rest.
+
+Advertisements are made available for consumption by indexer nodes as a set of
+files fetched via HTTP.
+
+Quoting from
+[Advertisement Transfer](https://github.com/ipni/specs/blob/90648bca4749ef912b2d18f221514bc26b5bef0a/IPNI.md#advertisement-transfer):
+
+> All IPNI HTTP requests use the IPNI URL path prefix, `/ipni/v1/ad/`. Indexers
+> and advertisement publishers implicitly use and expect this prefix to precede
+> the requested resource.
+>
+> The IPLD objects of advertisements and entries are represented as files named
+> by their CIDs in an HTTP directory. These files are immutable, so can be
+> safely cached or stored on CDNs. To fetch an advertisement or entries file by
+> CID, the request made by the indexer to the publisher is
+> `GET /ipni/v1/ad/{CID}`.
+
+The IPNI instance running at https://cid.contact provides an API returning the
+list of all index providers from which cid.contact have received announcements:
+
+https://cid.contact/providers
+
+The response provides all the metadata we need to download the advertisements:
+
+- `Publisher.Addrs` describes where we can contact SP's index provider
+- `LastAdvertisement` contains the CID of the head advertisement
+
+## Proposed Design
+
+### Ingestion
+
+Ingesting announcements from all Storage Providers is the most complex
+component. For each storage provider, we need to periodically check for the
+latest advertisement head and process the chain from head until we find an
+advertisement we have already processed before. The chain can be very long,
+therefore we need to account for the cases when the service restarts or a new
+head is published before we finish processing the chain.
+
+#### Proposed algorithm
+
+Use the following per-provider state persisted in the database:
+
+- `provider_id` - Primary key.
+
+- `provider_address` - Provider's address where we can fetch advertisements
+  from.
+
+- `last_head` - The CID of the head where we started the previous walk. All
+  advertisements from `last_head` to the end of the chain were already
+  processed.
+
+- `next_head` - The CID of the most recent head seen by cid.contact. This is
+  where we need to start the next walk from.
+
+- `head` - The CID of the head advertisement we started the current walk from.
+  We update this value whenever we start a new walk.
```

Comment:
Why do we need this state? Should it not be enough to go from next_head to last_head and afterwards update last_head?

Or is this for the case where this takes a long time and we want to be able to resume the chain walk? In this case, what about we simplify the design by saying that we will only ever walk X links per iteration, thereby eliminating the need for the `head` state?

Reply:
Then we could also remove `tail`

Reply:
The current walk starts from `head` and walks up to `last_head`. When the current walk reaches `last_head`, we need to set `last_head ← head` so that the next walk knows where to stop.

`next_head` is updated every minute when we query cid.contact for the latest heads. If the walk takes longer than a minute to finish, then `next_head` will change and we cannot use it for `last_head`.

What we can do, is to remove `next_head`, as I explained in https://github.com/filecoin-station/piece-indexer/pull/2/files#r1689865074

> In this case, what about we simplify the design by saying that we will only ever walk X links per iteration, thereby eliminating the need for the head state?

We must always walk the chain all the way to the genesis or to the entry we have already seen & processed. Here is how the state looks like in the middle of a walk:

```
next_head
  ↓
(entries announced after we started the current walk)
  ↓
head
  ↓
(entries visited in this walk)
  ↓
tail
  ↓
(entries NOT visited yet)
  ↓
last_head
  ↓
(entries visited in the previous walk)
  ↓
(null)
```